### PR TITLE
Autoplay Implementation

### LINKED ISSUE #164: Autoplay research and implementation
## Feature description

* Introduce methods of implementing video autoplay on sites.
* Implement autoplay on one of the videos.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by dero**
## Summary

* Adds a new video as a demonstration of how video autoplay can be integrated.
* Implements autoplay logic in JS.
* First, autoplay with sound is attempted.
* If that fails, autoplay without sound is attempted and unmute button is rendered.
* If that fails, a static play button is rendered.

### Screencast of the implemented feature

Notice how the video plays muted on refresh – simulating no domain interaction prior to arriving at the page with video – and how the "Unmute" button is rendered.

Also notice how the video plays with sound when user navigates to the video post within the site.

https://user-images.githubusercontent.com/433570/136047635-732ab64f-87b1-485c-9cef-0dcfc25e208c.mp4

### To-do

- [x] Thumbnail and Media Session artwork for the Autoplay video.

Fixes #164

### CODE REVIEW

**Comment by derekherman** [comment]
@dero just for clarity we should always branch off `develop` unless we are doing a hotfix to an existing release and then it would be off `main`, and probably shouldn't merge `main` into the branch. I was going to say something in the last PR but figured it was a one off thing.

Is the only reason you changed the folder structure of the API to sort them? Also, it should **not** be be called `Additional Features` this section is called `User Experience`—I believe we discussed that in Slack already.

**Comment by dero** [comment]
> just for clarity we should always branch off develop

@derekherman Yes, I realized this mid-way while working on this branch. Notice that I merged `main` 9 days ago and `develop` only recently.

I'm going to follow the right process from now on and I have checked that `develop` is a clean superset of `main` at this point, so no harm should be done.

> Is the only reason you changed the folder structure of the API to sort them?

The categories sections on the homepage are sorted alphabetically by their folder name, so the folder names matter.

I wanted to decouple category names from their position on the homepage. This was before we agreed on calling the new category `User Experience`.

> Also, it should not be be called Additional Features this section is called User Experience—I believe we discussed that in Slack already.

Right, I remember. 👍 Going to make the change today.

**Comment by derekherman** [comment]
Not sure why the linter isn’t running. We need to restructure the action because it doesn’t support external forks but that’s not the reason here.

File: src/api/03-user-experience/07-autoplay.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
+---
+title: Autoplay
+description: |
+  Learn how to start video playback automatically and employ autoplay strategies that don't lead to degraded user experience.
```
Comment (dero) [inline src/api/03-user-experience/07-autoplay.md:4]: @derekherman This text needs review.

File: src/api/03-user-experience/index.md [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+name: User Experience
+slug: user-experience
+---
+
+Unleash your videos and learn how autoplay, picture-in-picture and casting can be implemented in ways that improve user experience.
```
Comment (dero) [inline src/api/03-user-experience/index.md:6]: @derekherman This category description needs review.

File: src/js/web-components/video-player/VideoPlayer.js [hunk 3]
```diff
@@ -66,6 +66,22 @@ export default class extends HTMLElement {
     );
   }
 
+  /**
+   * Mutes the video.
+   */
+  mute() {
+    this.videoElement.setAttribute('muted', '');
```
Comment (beaufortfrancois) [inline src/js/web-components/video-player/VideoPlayer.js:73]: Why do you do both? HTML and JS.
I'd use only JS.

File: src/js/sw/cache.js [hunk 4]
```diff
@@ -43,4 +43,5 @@ export default [
   'https://storage.googleapis.com/kino-assets/streaming-basics/thumbnail.png',
   'https://storage.googleapis.com/kino-assets/efficient-formats/thumbnail.png',
   'https://storage.googleapis.com/kino-assets/adaptive-streaming/thumbnail.png',
+  'https://storage.googleapis.com/kino-assets/using-webvtt/thumbnail.png',
```
Comment (beaufortfrancois) [inline src/js/sw/cache.js:46]: Is that related to autoplay?

File: src/js/pages/Video.js [hunk 5]
```diff
@@ -18,104 +18,227 @@ import appendVideoToGallery from '../utils/appendVideoToGallery';
...
+  /**
+   * Sets up the manual play button.
+   */
+  const setupManualPlayButton = () => {
+    renderOverlay({
+      buttonElement: createOverlayButton({
+        markup: playIcon,
+        action: 'play',
+        callback: async () => {
+          await attachAndPlay();
```
Comment (beaufortfrancois) [inline src/js/pages/Video.js:193]: Is the `await` required there?

**Review by beaufortfrancois — COMMENTED** [review]
I've left some comments. Let me know when this PR is ready for review.

File: src/js/web-components/video-player/VideoPlayer.js [hunk 6]
```diff
@@ -285,18 +307,32 @@ export default class extends HTMLElement {
...
+        try {
+          await this.videoElement.play();
+
+          this.internal.downloader.download({
+            assetsOnly: true,
+          });
+
+          resolve(this);
+        } catch (_) {
+          reject(this);
```
[the remark this answers is not in the document]
Comment (dero) [reply src/js/web-components/video-player/VideoPlayer.js:328]: Added the comment in 4b2a4e5.

File: src/js/web-components/video-player/VideoPlayer.js [hunk 7]
```diff
@@ -285,18 +307,32 @@ export default class extends HTMLElement {
...
+    return new Promise((resolve, reject) => {
+      const resolvePlayIntent = async () => {
+        try {
+          await this.videoElement.play();
+
+          this.internal.downloader.download({
+            assetsOnly: true,
+          });
+
+          resolve(this);
```
[the remark this answers is not in the document]
Comment (dero) [reply src/js/web-components/video-player/VideoPlayer.js:326]: I added a [comment](https://github.com/GoogleChrome/kino/pull/165/files#diff-560daf41a4242202b45da3fe542f12d15aabd52fc81a05dee5a9da59bccaededR309) to the `@returns`  tag that explains why the promise resolves with a value.
